<a href="https://colab.research.google.com/github/innysik222/Pandas-Numpy-ML-/blob/main/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22practice_numpy_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://drive.google.com/uc?id=1_UPhD0m3AbkSKA6oy7oNqBfyJcar4lOC'>

In [ ]:
#@title Установка нужных версий библиотек
!wget 'https://drive.google.com/uc?export=download&id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn' -O requirements.txt
!pip install -r requirements.txt

--2022-05-23 13:33:10--  https://drive.google.com/uc?export=download&id=1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn
Resolving drive.google.com (drive.google.com)... 173.194.216.101, 173.194.216.113, 173.194.216.100, ...
Connecting to drive.google.com (drive.google.com)|173.194.216.101|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-0g-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/djbnqnsnsh3098mg1u5mpqi9neuq05g1/1653312750000/14904333240138417226/*/1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn?e=download [following]
--2022-05-23 13:33:10--  https://doc-0g-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/djbnqnsnsh3098mg1u5mpqi9neuq05g1/1653312750000/14904333240138417226/*/1oSFOP0j25OZAuhD8YXxyQXNTdr2lUdtn?e=download
Resolving doc-0g-c0-docs.googleusercontent.com (doc-0g-c0-docs.googleusercontent.com)... 172.217.193.132, 2607:f8b0:400c:c03::84
Connecting to doc-0g-c0-docs.googleusercontent.com (doc-0g-c0

## План занятия:
- Изучение набора данных с помощью атрибутов numpy массивов
- Применение математических операции к numpy массивам
- Индексирование
- Фильтрация данных и выбыр необходимых
- Итоги

## 1. Введение в numpy

### 1.1 Импортируйте библиотеку numpy и дайте ей псевдоним np 

In [ ]:
import numpy as np

### 1.2 Считайте данные из txt файлы

Файл находится на google диске: https://drive.google.com/file/d/1aT6L4dPtz0XaYXGpumwn2W1TpC6TIujl

В нем представлены данные людей:
- 1 колонка - возраст человека
- 2 колонка - вес человека
- 3 колонка - рост человека

In [ ]:
!wget 'https://drive.google.com/uc?id=1aT6L4dPtz0XaYXGpumwn2W1TpC6TIujl' -O data_practice.txt

--2022-05-23 13:33:13--  https://drive.google.com/uc?id=1aT6L4dPtz0XaYXGpumwn2W1TpC6TIujl
Resolving drive.google.com (drive.google.com)... 173.194.216.139, 173.194.216.101, 173.194.216.102, ...
Connecting to drive.google.com (drive.google.com)|173.194.216.139|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-0g-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/cjubuvk8353j0fgqvo8g6cgbi4dp2t9k/1653312750000/14904333240138417226/*/1aT6L4dPtz0XaYXGpumwn2W1TpC6TIujl [following]
--2022-05-23 13:33:14--  https://doc-0g-c0-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/cjubuvk8353j0fgqvo8g6cgbi4dp2t9k/1653312750000/14904333240138417226/*/1aT6L4dPtz0XaYXGpumwn2W1TpC6TIujl
Resolving doc-0g-c0-docs.googleusercontent.com (doc-0g-c0-docs.googleusercontent.com)... 172.217.193.132, 2607:f8b0:400c:c03::84
Connecting to doc-0g-c0-docs.googleusercontent.com (doc-0g-c0-docs.googleusercontent.com)|172.217.1

### 1.3. Выведите основную информацию по данным

1. Сколько есть людей в данных?
2. Сколько измерений есть в данных?
3. Какой тип данных?
4. Сколько всего есть элементов в данных?

In [ ]:
data = np.loadtxt('data_practice.txt')
data.shape


(100, 3)

In [ ]:
data.size

300

In [ ]:
data.dtype

dtype('float64')

## 2. Индексирование и математические операции

### 2.1 Посмотрите данные по первым 5 людям

In [ ]:
data[:5]

array([[ 43.,  62., 170.],
       [ 46.,  90., 173.],
       [ 35.,  54., 162.],
       [ 48., 104., 175.],
       [ 39.,  89., 166.]])

### 2.2 Найдите рост самого высокого человека

In [ ]:
data[:,2].max()

189.0

### 2.3 Выведите данные по самому молодому человека

In [ ]:
idx=data[:,0].argmin()
data[idx]

array([ 23.,  38., 150.])

### 2.4 Выведите средние значения веса и роста

In [ ]:
data[:,[1,2]].mean(axis=0)


array([ 78.94, 169.81])

### 2.5 Посчитайте ИМТ (BMI) людей

ИМТ - Индекс массы тела (англ. BMI - body mass index) - это величина, применяемая для оценки степени соответствия роста и массы человека и оценки массы тела (масса тела может быть нормальной, недостаточной, избыточной). Индекс массы тела рассчитывается по формуле: $$ИМТ = \frac{масса (кг)}{рост^2 (м^2)}$$

*Обратите внимание на измерения роста, они представлены в данных в сантиметрах.*

In [ ]:
#imt=np.array(data[:,1]/(data[:,2]/100)**2)
weight=data[:,1]
height=(data[:,2]/100)**2
imt=weight/height
imt

array([21.4532872 , 30.07116843, 20.57613169, 33.95918367, 32.29786616,
       33.26706966, 23.7953599 , 33.69005153, 34.14007572, 32.03125   ,
       31.70828379, 22.05805119, 22.38631428, 25.35154137, 21.484375  ,
       22.64737696, 30.48668503, 28.982007  , 32.97959184, 28.02768166,
       26.29172382, 24.22145329, 20.86111966, 27.34375   , 26.8809517 ,
       21.71806609, 30.45776225, 24.69040546, 24.78425752, 33.62208773,
       23.42355741, 27.08415907, 23.183391  , 21.05170826, 30.86300683,
       22.92097107, 26.30361892, 25.53544639, 30.47796622, 28.73192149,
       32.421875  , 21.79944511, 22.15102251, 31.1613739 , 34.36791383,
       27.44059917, 31.91222346, 25.9043101 , 21.2585034 , 31.63739669,
       22.5981405 , 28.72008325, 32.1799308 , 21.10726644, 27.09925101,
       30.43671557, 24.03460984, 30.05681068, 34.02034615, 33.46403461,
       29.95436858, 28.73174689, 21.4532872 , 27.88518739, 28.05570579,
       30.48315806, 30.86419753, 24.9107674 , 28.04281659, 27.17

## 3. Фильтрация по данным

### 3.1 Выведите только данные тех людей, у которых ИМТ находится в категории "Ожирение (Класс I)" и выше.

Категории ИМТ:

- Крайний недостаток веса - менее 16.5 
- Недостаток в весе от 16.5 до 18.4 
- Нормальный вес тела от 18.5 до 24.9 
- Избыточная масса тела от 25 до 30
- Ожирение (Класс I) от 30.1 до 34.9 
- Ожирение (Класс II - тяжелое) от 35 до 40 
- Ожирение (Класс III - крайне тяжелое) более 40




In [ ]:
imt_big=data[imt>=30.1]
imt_big

array([[ 48., 104., 175.],
       [ 39.,  89., 166.],
       [ 30.,  82., 157.],
       [ 47., 102., 174.],
       [ 45., 101., 172.],
       [ 33.,  82., 160.],
       [ 47.,  96., 174.],
       [ 38.,  83., 165.],
       [ 48., 101., 175.],
       [ 56., 102., 183.],
       [ 32.,  85., 159.],
       [ 34.,  80., 161.],
       [ 40.,  85., 167.],
       [ 33.,  83., 160.],
       [ 42.,  89., 169.],
       [ 41.,  97., 168.],
       [ 40.,  89., 167.],
       [ 49.,  98., 176.],
       [ 43.,  93., 170.],
       [ 44.,  89., 171.],
       [ 47., 103., 174.],
       [ 45.,  99., 172.],
       [ 35.,  80., 162.],
       [ 53., 100., 180.],
       [ 52., 105., 179.],
       [ 54., 102., 181.],
       [ 49.,  99., 176.],
       [ 34.,  80., 161.],
       [ 38.,  87., 165.],
       [ 30.,  77., 157.],
       [ 41.,  95., 168.],
       [ 50., 100., 177.],
       [ 47.,  95., 174.],
       [ 45.,  98., 172.]])

### 3.2 Найдите в этой группе людей средний возраст, вес и рост

In [ ]:
# Ваш код здесь
imt_big_mean=imt_big.mean(axis=0)
imt_big_mean

array([ 42.61764706,  92.64705882, 169.61764706])

### 3.3 Пометьте в исходном наборе данных людей с п.1 единицами

In [ ]:
#@title Подсказки
# 1. Создайте заготовку состоящую из нулей
# 2. Просоедините эту заготовку к исходному набору данных
# 3. Проставьте единицы в новый столбик, если BMI > 30.1

In [ ]:
# Ваш код здесь
imt_excessive=np.zeros(100)
imt_excessive
data=np.column_stack([data,imt_excessive])

data[imt>=30.1,3]=1
data

array([[ 43.,  62., 170.,   0.],
       [ 46.,  90., 173.,   0.],
       [ 35.,  54., 162.,   0.],
       [ 48., 104., 175.,   1.],
       [ 39.,  89., 166.,   1.],
       [ 30.,  82., 157.,   1.],
       [ 37.,  64., 164.,   0.],
       [ 47., 102., 174.,   1.],
       [ 45., 101., 172.,   1.],
       [ 33.,  82., 160.,   1.],
       [ 47.,  96., 174.,   1.],
       [ 42.,  63., 169.,   0.],
       [ 46.,  67., 173.,   0.],
       [ 45.,  75., 172.,   0.],
       [ 33.,  55., 160.,   0.],
       [ 45.,  67., 172.,   0.],
       [ 38.,  83., 165.,   1.],
       [ 55.,  96., 182.,   0.],
       [ 48., 101., 175.,   1.],
       [ 43.,  81., 170.,   0.],
       [ 35.,  69., 162.,   0.],
       [ 43.,  70., 170.,   0.],
       [ 44.,  61., 171.,   0.],
       [ 33.,  70., 160.,   0.],
       [ 60.,  94., 187.,   0.],
       [ 46.,  65., 173.,   0.],
       [ 56., 102., 183.,   1.],
       [ 34.,  64., 161.,   0.],
       [ 56.,  83., 183.,   0.],
       [ 32.,  85., 159.,   1.],
       [ 3

### 3.4 Проверьте правильность простановки единиц в новой колонке

Возьмите одного любого человека с единицей в новом столбце и посчитайте его bmi.

In [ ]:
# Ваш код здесь
imt[5]

33.26706965799829

### 3.5 Посчитайте, сколько людей находится в каждой группе по ИМТ


Категории ИМТ:

- Крайний недостаток веса - менее 16.5 
- Недостаток в весе от 16.5 до 18.4 
- Нормальный вес тела от 18.5 до 24.9 
- Избыточная масса тела от 25 до 30
- Ожирение (Класс I) от 30.1 до 34.9 
- Ожирение (Класс II - тяжелое) от 35 до 40 
- Ожирение (Класс III - крайне тяжелое) более 40

*Объединять условия в numpy можно через `&`, через более привычный `and`, к сожалению, не получится*

In [ ]:
# Ваш код здесь
imt_1=data[imt<16.5].shape[0]

imt_2=data[(imt>=16.5) & (imt<=18.4)].shape[0]
imt_2


1

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

### 3.6 Убедитесь, что суммарно по категориям людей вышло 100

In [ ]:
# Ваш код здесь

#### 3.6.1 Если людей вышло меньше, то подумайте, почему так вышло

In [ ]:
# Ваши мысли здесь

#### 3.6.2 И как это можно исправить

In [ ]:
# Ваш код здесь

#### 3.6.3 Заново посчитайте количество людей в каждой категории

In [ ]:
# Ваш код здесь

#### 3.6.4 И проверьте, что суммарно вышло 100 людей

In [ ]:
# Ваш код здесь

## Резюме
Сегодня на практическом занятии:
1. Поизучали набор данных с помощью атрибутов numpy массивов
2. Применили математические операции к numpy массивам
3. Пофильтровали данные и выбрали для нас необходимые